services-core/src/solution_submission/retry.rs

```diff
@@ -399,26 +414,61 @@ mod tests {
             .times(1)
             .return_once(|_, _, _, _, _| {
                 async move {
-                    receiver.await.unwrap();
+                    submit_receiver.await.unwrap();
                     Ok(())
                 }
                 .boxed()
             });
-        sleep.expect_sleep().times(1).returning(|_| immediate!(()));
+        sleep.expect_sleep().return_once(move |_| {
```

Comment:
I'm not a big fan of using the sleep method here to kickstart the future resolution as it relies on pretty detailed knowledge of the intrinsics (e.g. the fact that sleep is called after the first solution submission).

Would it work to `cancel_sender.send(()).unwrap();` inside `expect_submit_solution` before `submit_receiver.await.unwrap()`? This way the test would say: we cancel as soon as we have issued the first tx and resolve the first tx as soon as we have started the cancellation.

Reply:
Good point, changed.